numpy/f2py/rules.py

```diff
@@ -215,6 +215,7 @@
 \td = PyModule_GetDict(m);
 \ts = PyString_FromString(\"$R""" + """evision: $\");
 \tPyDict_SetItemString(d, \"__version__\", s);
+\tPy_DECREF(s);
```

Comment:
This is the source of the double free on Python3.5. When I remove it I still get a segfault, but not a `double free` and abort

Reply:
Seems strange that this should be so bad, it seems obviously right, no? Don't really have time to look into it though.

Reply:
I have observed double free with Python 3.5 as well, and the fix (counter-intuitively) is to use `PyObject_Del` instead of `PyMem_Free` (see gh-14217).

Reply:
Hmmm,` PyObject_Free` is the correct thing in any case I think. Although, I am a bit puzzle why it ends up crashing things so hard, but then I do not know how the python memory management works internally. I think I will rebase your `PyObject_Free` fix as a first commit into my PR?

Reply:
Sgtm!